### ISSUE TITLE
Cannot locate static files

### ISSUE THREAD

**Issue by convers39** [testdrivenio/django-on-docker-letsencrypt#7]
Hi there,
First of all, thank you so much for this tutorial, your site and tutorials are great, it really helps me. 
I am learning for a few months and not familiar with docker or deployment, I followed your steps and it was almost down now.
While I have an issue that cannot figure out why the static files are not served as expected. And wonder if you could have a look at my issue, really really appreciate it.

### static_url and static_root
 I was using `static` as url and root before but didnt work, now change to the same as the tutorial.
```python
# settings/prod.py
STATIC_URL = '/staticfiles/'
STATIC_ROOT =  os.path.join(BASE_DIR, 'staticfiles')
```
Inside a Django app shell `settings.STATIC_ROOT` will return `'/home/app/web/staticfiles'`, I believe this should be what we expected?

### dockerfile for production
Here I didn't use the alpine version as too many issues on dependencies for me to handle. And also I comment out the shell exec file since it cannot be executed for permission issues even I add exec permission. Will those changes influence serving static files?
```yml
###########
# BUILDER #
###########

# pull official base image
FROM python:3.8.6-slim-buster as builder

# set work directory
WORKDIR /usr/src/app

# set environment variables
ENV PYTHONDONTWRITEBYTECODE 1
ENV PYTHONUNBUFFERED 1

# install psycopg2 dependencies
# RUN apk update && \
#     apt-get upgrade -y && \
#     apk add postgresql-dev gcc python3-dev musl-dev && \
#     apt-get clean
RUN apt-get update && \
    apt-get upgrade -y && \
    apt-get install -y netcat-openbsd gcc && \
    apt-get clean

# lint
RUN pip install --upgrade pip

COPY . .
# install dependencies
COPY ./requirements.txt .
RUN pip wheel --no-cache-dir --no-deps --wheel-dir /usr/src/app/wheels -r requirements.txt


#########
# FINAL #
#########

# pull official base image
FROM python:3.8.6-slim-buster

# create directory for the app user
RUN mkdir -p /home/app

# create the app user
# RUN addgroup -S app && adduser -S app -G app
RUN addgroup --system admin && adduser --system --group admin

# create the appropriate directories
ENV HOME=/home/app
ENV APP_HOME=/home/app/web
RUN mkdir $APP_HOME
RUN mkdir $APP_HOME/staticfiles
RUN mkdir $APP_HOME/media
WORKDIR $APP_HOME

# install dependencies
# RUN apt-get update && apt-get install -y libpq
COPY --from=builder /usr/src/app/wheels /wheels
COPY --from=builder /usr/src/app/requirements.txt .
RUN pip install --no-cache /wheels/*

# copy entrypoint-prod.sh
# COPY ./entrypoint.prod.sh $APP_HOME

# copy project
COPY . $APP_HOME

# chown all the files to the app user
RUN chown -R admin:admin $APP_HOME

# change to the app user
USER admin

# run entrypoint.prod.sh
# ENTRYPOINT ["/home/app/web/entrypoint.prod.sh"]
```

### nginx/vhost setting file
Same as the tutorial
```
location /staticfiles/ {
  alias /home/app/web/staticfiles/;
  add_header Access-Control-Allow-Origin *;
}
```
### nginx dockerfile
Same as the tutorial.
```yml
FROM jwilder/nginx-proxy
COPY vhost.d/default /etc/nginx/vhost.d/default
COPY custom.conf /etc/nginx/conf.d/custom.conf
```
### docker-compose file. 
I didn't create an app folder and Nginx folder, instead, the Nginx folder is in my Django app root directory.
```yml
version: "3.8"

services:
  django:
    restart: always
    build:
      context: .
      dockerfile: ./dockerfile.prod
    container_name: django
    image: 666666666.dkr.ecr.mmmmmm.amazonaws.com/django-ec2:django
    command: gunicorn --workers=2 --bind=0.0.0.0:8000 --env DJANGO_SETTINGS_MODULE=core.settings.prod core.wsgi:application
    volumes:
      - static_volume:/home/app/web/staticfiles
      - media_volume:/home/app/web/media
    expose:
      - 8000
    env_file:
      - ./core/env/env.prod
    depends_on:
      - redis
  celery:
    restart: always
    build:
      context: .
      dockerfile: ./dockerfile.prod
    container_name: celery
    entrypoint: /bin/sh
    command: ./docker.celery.sh
    volumes:
      - static_volume:/home/app/web/staticfiles
      - media_volume:/home/app/web/media
    env_file:
      - ./core/env/env.prod
    depends_on:
      - django
      - redis
  redis:
    restart: always
    image: redis:alpine
    container_name: redis
    volumes:
      - redisdata:/data
  nginx-proxy:
    container_name: nginx-proxy
    build: ./nginx
    image: 66666666.dkr.ecr.mmmmmmm.amazonaws.com/django-ec2:nginx
    restart: always
    ports:
      - 443:443
      - 80:80
    volumes:
      - static_volume:/home/app/web/staticfiles
      - media_volume:/home/app/web/media
      - certs:/etc/nginx/certs
      - html:/usr/share/nginx/html
      - vhost:/etc/nginx/vhost.d
      - /var/run/docker.sock:/tmp/docker.sock:ro
    depends_on:
      - django
  nginx-proxy-letsencrypt:
    image: jrcs/letsencrypt-nginx-proxy-companion
    env_file:
      - ./core/env/env.prod.proxy-companion
    volumes:
      - /var/run/docker.sock:/var/run/docker.sock:ro
      - certs:/etc/nginx/certs
      - html:/usr/share/nginx/html
      - vhost:/etc/nginx/vhost.d
    depends_on:
      - nginx-proxy
volumes:
  redisdata:
  static_volume:
  media_volume:
  certs:
  html:
  vhost:

```

### my file structure:
```sh
.
├── LICENSE
├── README.md
├── apps
├── core
│   ├── __init__.py
│   ├── __pycache__
│   ├── asgi.py
│   ├── celery.py
│   ├── env
│   ├── settings
│   ├── urls.py
│   └── wsgi.py
├── docker-compose.prod.yml
├── docker-compose.staging.yml
├── docker-compose.yml
├── docker.celery.sh
├── docker.django.sh
├── dockerfile
├── dockerfile.prod
├── irecommerce.code-workspace
├── manage.py
├── media
├── nginx
│   ├── certs
│   ├── custom.conf
│   ├── default.conf
│   ├── dockerfile
│   └── vhost.d
├── requirements.txt
├── static
└── templates
```
### Issue
Now when I compose up the containers, the ssl certs are issued properly, and it is possible to visit the domain, while all static files are not served and I got this error message in the terminal:
```sh
nginx-proxy                | nginx.1    | 2020/12/28 05:12:52 [error] 88#88: *1 open()
 "/usr/src/app/static/js/front.js" failed (2: No such file or directory), client: 66666ip, 
server: domain.xyz, request: "GET /staticfiles/js/front.js HTTP/2.0", host: "domain.xyz", 
referrer: "https://domain.xyz/"
```
```sh
"GET /staticfiles/js/front.js HTTP/2.0" 404 555 "https://domain.xyz/"
```
It seems trying to find static files in `/usr/src/app/static/` folder, thus I don't understand why this is happening since there are no settings refer to this directory.

I am not sure if the information above is enough, it would be really appreciated if I could have some advice..

### Some extra info:
If I enter the django app container, the file looks like this:
```sh
app@3f71d7307cde:~/web$ pwd
/home/app/web
app@3f71d7307cde:~/web$ ls
LICENSE                  docker-compose.staging.yml  manage.py
README.md                docker-compose.yml          media
_samples                 docker.celery.sh            nginx
apps                     docker.django.sh            requirements.txt
celerybeat-schedule      dockerfile                  static
core                     dockerfile.prod             staticfiles
db                       entrypoint.prod.sh          templates
docker-compose.prod.yml  irecommerce.code-workspace
```
I have run `collectstatic` and all necessary static files are in `staticfiles` folder, this should be consistent with the settings in `nginx/vhost/default`.

**Comment by convers39** [comment]
**Updata**
Somehow I fixed this error:
```shell
nginx-proxy                | nginx.1    | 2020/12/28 05:12:52 [error] 88#88: *1 open()
 "/usr/src/app/static/js/front.js" failed (2: No such file or directory), client: 66666ip, 
server: domain.xyz, request: "GET /staticfiles/js/front.js HTTP/2.0", host: "domain.xyz", 
referrer: "https://domain.xyz/"
```
I changed all my staticfile dir settings back to `static` only, instead of `staticfiles`, and in the container I use `usr/src/app` as the working directory instead of making a home dir for the user. I  guess there is some inconsistency in the directory which leads to the error above.

However, the static files are still not working properly. Both django and nginx-proxy containters return this 404 message.

```shell
django                     | Not Found: /static/js/front.js
nginx-proxy                | nginx.1    | domain.xyz x.8x2.2x2.66 - - [28/Dec/2020:14:26:26 +0000] "GET /static/js/front.js HTTP/2.0" 404 10638 "https://domain.xyz/" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/87.0.4280.88 Safari/537.36"
```

I recheck the tutorial several times, as my understanding, the static file's routing config was created by `nginx-proxy` automatically, thus we didnt create any `default.conf` file to `nginx-proxy`, but only the setting in `vhost.d` as the router for static files?

https://testdriven.io/blog/django-lets-encrypt/
> As mentioned, VIRTUAL_HOST (and VIRTUAL_PORT) are needed by nginx-proxy to auto create the reverse proxy configuration.
> Requests that match any of these patterns will be served from static or media folders. They won't be proxied to other containers. The web and nginx-proxy containers share the volumes in which the static and media files are located:
```shell
location /staticfiles/ {
  alias /home/app/web/staticfiles/;
  add_header Access-Control-Allow-Origin *;
}
```

**Comment by convers39** [comment]
In case just update my current settings, changed the dir, and use my previous dockerfile.
Static files are in `/usr/src/app/static` folder, already run `collectstatic`.

static root and url:
```python
STATIC_URL = '/static/'
STATIC_ROOT = os.path.join(BASE_DIR, 'static')
```


dockerfile:

```dockerfile
FROM python:3.8.6-slim-buster

RUN apt-get update && \
    apt-get upgrade -y && \
    apt-get install -y netcat-openbsd gcc && \
    apt-get clean

ENV PYTHONUNBUFFERED=1

WORKDIR /usr/src/app

RUN addgroup --system admin && adduser --system --no-create-home --group admin && \
    chown -R admin:admin /usr && chmod -R 755 /usr

COPY ./requirements.txt /usr/src/app/requirements.txt
RUN pip install --upgrade pip \
    pip install -r requirements.txt

USER admin

COPY . /usr/src/app
```

docker-compose.prod:

```yml
version: "3.8"

services:
  django:
    restart: always
    build:
      context: .
      dockerfile: ./dockerfile
    container_name: django
    command: gunicorn --workers=2 --bind=0.0.0.0:8000 --env DJANGO_SETTINGS_MODULE=core.settings.prod core.wsgi:application
    volumes:
      - .:/usr/src/app
      - static_volume:/usr/src/app/static
      - media_volume:/usr/src/app/media
      # - ./nginx/certs/:/etc/certs
    ports:
      - "8000:8000"
    env_file:
      - ./core/env/env.prod
    depends_on:
      - redis
  celery:
    restart: always
    build:
      context: .
      dockerfile: ./dockerfile
    container_name: celery
    entrypoint: /bin/sh
    command: ./docker.celery.sh
    volumes:
      - .:/usr/src/app
      - static_volume:/usr/src/app/static
      - media_volume:/usr/src/app/media
    env_file:
      - ./core/env/env.prod
    depends_on:
      - django
      - redis
  redis:
    restart: always
    image: redis:alpine
    container_name: redis
    volumes:
      - redisdata:/data
  nginx-proxy:
    container_name: nginx-proxy
    build: ./nginx
    restart: always
    ports:
      - 443:443
      - 80:80
    volumes:
      - static_volume:/usr/src/app/static
      - media_volume:/usr/src/app/media
      - certs:/etc/nginx/certs
      - html:/usr/share/nginx/html
      - vhost:/etc/nginx/vhost.d
      - /var/run/docker.sock:/tmp/docker.sock:ro
    depends_on:
      - django
  nginx-proxy-letsencrypt:
    image: jrcs/letsencrypt-nginx-proxy-companion
    env_file:
      - ./core/env/env.prod.proxy-companion
    volumes:
      - /var/run/docker.sock:/var/run/docker.sock:ro
      - certs:/etc/nginx/certs
      - html:/usr/share/nginx/html
      - vhost:/etc/nginx/vhost.d
    depends_on:
      - nginx-proxy
volumes:
  redisdata:
  static_volume:
  media_volume:
  certs:
  html:
  vhost:

```



nginx dockerfile:

```dockerfile
FROM jwilder/nginx-proxy
COPY vhost.d/default /etc/nginx/vhost.d/default
COPY custom.conf /etc/nginx/conf.d/custom.conf
```



nginx/vhost.d/default

```nginx
location /static/ {
  alias /usr/src/app/static/;
  add_header Access-Control-Allow-Origin *;
}
```



file tree:

```shell
.
├── LICENSE
├── README.md
├── apps
├── core
│   ├── __init__.py
│   ├── __pycache__
│   ├── asgi.py
│   ├── celery.py
│   ├── env
│   ├── settings
│   ├── urls.py
│   └── wsgi.py
├── docker-compose.prod.yml
├── docker-compose.staging.yml
├── docker-compose.yml
├── docker.celery.sh
├── docker.django.sh
├── dockerfile
├── dockerfile.prod
├── irecommerce.code-workspace
├── manage.py
├── media
├── nginx
│   ├── certs
│   ├── custom.conf
│   ├── default.conf
│   ├── dockerfile
│   └── vhost.d
├── requirements.txt
├── static
└── templates
```

**Comment by convers39** [comment]
**Update SOLVED**
Well actually still not very clear why but here is the thing:
I looked into the `nginx-proxy` container and found that the routing of static files didnt change to `/static/` after I change the local file `/nginx/vhost.d/default`.
After I changed `default` in the container manually, all solved...

The unclear part for me is that it seems when I rebuild the `nginx-proxy` image, which will run this copy instruction, it did not recopy the `default` file into the container after I changed the `default` file? 
In other words, does it mean that docker kept using the cached layer of `COPY vhost.d/default /etc/nginx/vhost.d/default` which holds an unchanged setting and different from my local changed files?
Or should I build a host volume say `.nginx/vhost.d/: /etc/nginx/vhost.d/` instead of using named volume `vhost: /etc/nginx/vhost.d/`？

```yml
# nginx/dockerfile
FROM jwilder/nginx-proxy
COPY vhost.d/default /etc/nginx/vhost.d/default
COPY custom.conf /etc/nginx/conf.d/custom.conf
```

**Comment by LennyLip** [comment]
`collectstatic` on local host is not good idea, just add

```
python manage.py migrate
python manage.py collectstatic
```

to [entrypoint.prod.sh ](https://github.com/testdrivenio/django-on-docker-letsencrypt/blob/master/app/entrypoint.prod.sh)
( after PostgreSQL started block)
